**One-bin A + B <-> C, with 1st-order kinetics for each species,
taken to equilibrium**

LAST REVISED: June 13, 2022

In [1]:
from modules.chemicals.chemicals import Chemicals as chem
from modules.reactions.reactions import Reactions
from life_1D.bio_sim_1d import BioSim1D as bio

from modules.html_log.html_log import HtmlLog as log

import set_path
set_path.add_ancestor_dir_to_syspath(3)  # The number of levels to go up 
                                         # to reach the project's home from the folder containing this notebook

Added 'D:\Docs\- MY CODE\BioSimulations\life123-Win7' to sys.path


In [2]:
# Initialize the system
chem_data = chem(diffusion_rates=[0.1, 0.1, 0.1], names=["A", "B", "C"])    # NOTE: diffusion_rates not used for now

rxn = Reactions(chem_data)

# Reaction A + B <-> C , with 1st-order kinetics for each species
rxn.add_reaction(reactants=[("A") , ("B")], products=[("C")],
                 forward_rate=5., reverse_rate=2.)

bio.initialize_system(n_bins=1, chem_data=chem_data, reactions=rxn)

bio.set_uniform_concentration(species_index=0, conc=10.)
bio.set_uniform_concentration(species_index=1, conc=50.)
bio.set_uniform_concentration(species_index=2, conc=20.)

bio.describe_state()

1 bins and 3 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [10.]
  Species 1 (B). Diff rate: 0.1. Conc:  [50.]
  Species 2 (C). Diff rate: 0.1. Conc:  [20.]


In [3]:
rxn.describe_reactions()

Number of reactions:  1
0: A + B <-> C  (Rf = 5.0 / Rb = 2.0)


In [4]:
rxn._internal_reactions_data()    # Low-level view of the reactions data

0: [(1, 0, 1), (1, 1, 1)] <-> [(1, 2, 1)]   ; Fwd: 5.0 / Back: 2.0


In [5]:
# First step
bio.react(time_step=0.002, n_steps=1)
bio.describe_state()

1 bins and 3 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [5.08]
  Species 1 (B). Diff rate: 0.1. Conc:  [45.08]
  Species 2 (C). Diff rate: 0.1. Conc:  [24.92]


_Early in the reaction :_
[A] = 5.08 ,  [B] = 45.08 ,  [C] = [24.92]

In [6]:
# Numerous more steps
bio.react(time_step=0.002, n_steps=40)

bio.describe_state()

1 bins and 3 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [0.29487741]
  Species 1 (B). Diff rate: 0.1. Conc:  [40.29487741]
  Species 2 (C). Diff rate: 0.1. Conc:  [29.70512259]


Consistent with the 5/2 ratio of forward/reverse rates (and the 1st order reactions),
the systems settles in the following equilibrium:  
[A] = 0.29487741 , [B] = 40.29487741 , [C] = [29.70512259]

# Note: "A" is largely the limiting reagent

In [7]:
A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
C_eq = bio.bin_concentration(0, 2)
print(f"Ratio of equilibrium concentrations (C_eq / (A_eq * B_eq)) : {C_eq / (A_eq * B_eq)}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations (C_eq / (A_eq * B_eq)) : 2.4999999765280356
Ratio of forward/reverse rates: 2.5
